tensorflow/python/ops/check_ops.py

```diff
@@ -1504,6 +1505,206 @@ def assert_type(tensor, tf_type, message=None, name=None):
+      if not isinstance(shape, (tuple, list)):
+        return tuple([shape])
+      return shape
+
+    def has_known_value(size):
+      return size is not None and isinstance(size, int)
+
+    actual_sizes_by_tensor = {
+        tensor: actual_dimension_sizes(tensor)
+        for tensor in shapes
+    }
+    specified_symbolic_sizes_by_tensor = {
+        tensor: specified_symbolic_dimension_sizes(shapes[tensor])
+        for tensor in shapes
+    }
+
+    size_assertions = []
+    size_specifications = {}
+    for x in shapes.keys():
+      actual_sizes = actual_sizes_by_tensor[x]
+      symbolic_sizes = specified_symbolic_sizes_by_tensor[x]
+
+      for i, size_symbol in enumerate(symbolic_sizes):
+
+        if size_symbol is None:
+          # Size specified with None implies no constraint
+          continue
+
+        if event_shape_only:
+          tensor_dim = i - len(symbolic_sizes)
+        else:
+          tensor_dim = i
+
+        if size_symbol in size_specifications or has_known_value(size_symbol):
+          if has_known_value(size_symbol):
+            specified_size = size_symbol
+            size_check_message = 'Specified explicitly'
+          else:
+            specified_size, specified_by_y, specified_at_dim = \
+                size_specifications[size_symbol]
+            if executing_eagerly:
+              name_y = _shape_and_dtype_str(specified_by_y)
+            else:
+              name_y = specified_by_y.name
+            size_check_message = 'Specified by tensor %s dimension %d' % \
+                                 (name_y, specified_at_dim)
+
+          if executing_eagerly:
+            name = _shape_and_dtype_str(x)
+          else:
+            name = x.name
+
+          actual_size = actual_sizes[tensor_dim]
+          if has_known_value(actual_size) and has_known_value(specified_size):
+            if actual_size != specified_size:
+              raise ValueError(
+                  '%s.  %s.  Tensor %s dimension %s must have size %d.  '
+                  'Received size %d, shape %s' %
+                  (message, size_check_message, name, tensor_dim,
+                   specified_size, actual_size, x.get_shape()))
+            # No dynamic assertion needed
+            continue
+
+          condition = math_ops.equal(
+              ops.convert_to_tensor(actual_size),
+              ops.convert_to_tensor(specified_size))
+          data_ = data
+          if data is None:
+            data_ = [
+                message,
+                size_check_message,
+                'Tensor %s dimension' % name, tensor_dim, 'must have size',
+                specified_size, 'Received shape: ', array_ops.shape(x)
+            ]
+          size_assertions.append(
+              control_flow_ops.Assert(condition, data_, summarize=summarize))
+        else:
+          size = actual_sizes[tensor_dim]
+          size_specifications[size_symbol] = (size, x, tensor_dim)
+
+    if not rank_assertions:
```

Comment:
Do you need this? I think tf.control_dependencies is perfectly happy to take an empty list.